File: Sorting/quickSort.py
def quickSort(array, left, right):
    if left < right: #This helps to check if there is still  or more elements in the sub-array. If not, then the element is in its correct position in the whole array and no need to sort.
        pivot = right
        partitionIndex = partition(array, pivot, left, right)

        quickSort(array, left, partitionIndex - 1) #left array
        quickSort(array, partitionIndex + 1, right) #right array
    
    return array  

def partition(array, pivot, left, right):
    pivotValue = array[pivot]
    partitionIndex = left 

    index = left

    while index < right:
        if array[index] < pivotValue:
            swap(array, index, partitionIndex)
            partitionIndex += 1  #to get the index where we will swap our pivot value into
        index += 1
    
    swap(array, right, partitionIndex) #here we swp the pivot value and value at the point where the pivot value should correctly be
    return partitionIndex

def swap(array, firstIndex, secondIndex):
    array[firstIndex],array[secondIndex]  = array[secondIndex], array[firstIndex]
    

numbers = [99,44,6,2,1,5,63,87,283,4,0]
quickSort(numbers, 0, len(numbers) - 1)

print(numbers)

File: Java Implementations/src/com/dataStructuresSorting/QuickSort.java
package com.dataStructuresSorting;

public class QuickSort {
    public void sort(int[] array){
        sort ( array, 0, array.length - 1 );
    }

    private void sort(int[] array, int start, int end){
        if (start >= end)
            return;

        var boundary = partition ( array, start, end );

        sort ( array, start, boundary - 1 );
        sort ( array, boundary + 1, end );
    }

    private int partition(int[] array, int start, int end){
        var pivot = array[end];
        var boundary = start - 1;
        for (int i = start; i <= end; i++)
            if (array[i] <= pivot)
                swap(array, i, ++boundary);

        return boundary;
    }

    private void swap ( int[] array, int firstIndex, int secondIndex ) {
        var temp = array[secondIndex];
        array[secondIndex] = array[firstIndex];
        array[firstIndex] = temp;
    }
}
